In [11]:
# All important libraries goes here!
import pandas as pd
import math

In [12]:
# Data Frame and removing colums
dataframe = pd.read_csv('./Data/washedData.csv')
dataframe = dataframe.drop(columns='ID')

#### Problem Statement
We are trying to understand the factors that influence whether a company has affected employees. This is important because companies with affected employees may require additional support or interventions. By identifying the key factors, we can target our interventions more effectively and potentially prevent employees from being affected in the future.

To solve this problem, we will use this dataset to build a predictive model. This model will take as input the various financial and operational characteristics of a company and output a prediction of whether the company has affected employees. We can then use this model to predict the status of new companies and guide our interventions.


<br>
<br>

#### (a) Data cleaning

* Handling duplicate values.
* Dealing with missing/null values.
* Addressing inconsistent data.

We will go through every column

<br>

#### (a) (ii) looking if there are missing values in each column of the dataframe

In [13]:
dataframe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4137 entries, 0 to 4136
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   No of employee     4137 non-null   int64 
 1   Annual turnover    4129 non-null   object
 2   TCTC               4123 non-null   object
 3   Basic Salary       4135 non-null   object
 4   Cash Injection     4137 non-null   int64 
 5   Contrib Waiver     4137 non-null   int64 
 6   Affected Employee  4137 non-null   int64 
dtypes: int64(4), object(3)
memory usage: 226.4+ KB


From the results it shows that there arent any missing values, since the total entries = Not-Null Count of 4137

<br>

#### (a) (iii) Checking for duplicate rows

In [14]:
dataframe.duplicated().sum()

1075

There are no duplicates in the dataset, for the sum = 0

<br>

#### (a) (iv) Ensuring data consistency

These attributes may contain some inconsistencies, for the datatype has to be an integer.
we will remove those values and run an algorithm to ensure consistency.

 2.   Annual turnover    4129 non-null   object
 3.   TCTC               4123 non-null   object
 4.  Basic Salary       4135 non-null   object

In [15]:
dataframe['Annual turnover'] = pd.to_numeric(dataframe['Annual turnover'], errors='coerce')
dataframe['TCTC'] = pd.to_numeric(dataframe['TCTC'], errors='coerce')
dataframe['Basic Salary'] = pd.to_numeric(dataframe['Basic Salary'], errors='coerce')

dataframe[['Annual turnover', 'TCTC', 'Basic Salary']].describe()

Annual turnover          TCTC  Basic Salary
count     3.876000e+03  3.978000e+03  4.099000e+03
mean      7.198401e+06  4.947246e+05  4.264880e+05
std       8.427608e+07  1.464206e+07  1.022407e+07
min      -3.968471e+08  0.000000e+00  0.000000e+00
25%       0.000000e+00  0.000000e+00  0.000000e+00
50%       4.000000e+04  3.600000e+03  5.000000e+03
75%       9.536879e+05  5.791195e+04  4.724824e+04
max       3.407609e+09  9.071940e+08  5.854672e+08

Some minor cleaning has been performed, and unknown values have been removed, For further cleaning, data will be formatted to readable format/values

In [16]:
dataframe['Annual turnover'] = dataframe['Annual turnover'].apply(lambda x: int(float(x)) if not math.isnan((x)) else '')
dataframe['TCTC'] = dataframe['TCTC'].apply(lambda x: int(float(x)) if not math.isnan((x)) else '')
dataframe['Basic Salary'] = dataframe['Basic Salary'].apply(lambda x: int(float(x)) if not math.isnan((x)) else '')

In [22]:
dataframe

No of employee Annual turnover      TCTC Basic Salary  Cash Injection  \
0                 63      3098000000  13782989      9500520               0   
1                 73        15000000    326574       992400               1   
2                 18               0    543629       397789               0   
3                 25        52000000    725607       496910               0   
4                  1          765270    205385        31530               0   
...              ...             ...       ...          ...             ...   
4132               1               0         0            0               0   
4133               2          550992     45916        45916               1   
4134               1               0         0            0               0   
4135               1               0         0            0               0   
4136               2         2500000         0            0               1   

      Contrib Waiver  Affected Employee  
0                  1                  0  
1                  1                  1  
2                  1                  1  
3                  1                  1  
4                  0                  0  
...              ...                ...  
4132               0                  1  
4133               1                  1  
4134               0                  1  
4135               0                  1  
4136               1                  1  

[4137 rows x 7 columns]

<br>
<br>

#### Statistical Analysis

* Mean TCTC: Explain the average compensation.
* Median Basic Salary: Discuss the central tendency.
* Standard Deviation of Annual Turnover: Highlight variability.

In [18]:
# Code goes here!

<br>
<br>

#### Exploratory Data Analysis (EDA)

* Histograms of "Annual turnover" and "Basic Salary."
* Box plots of "TCTC" to identify outliers.
* Scatter plots to explore relationships between variables.

Interpret the insights gained from each visualization.

In [19]:
# Code goes here!
# Ever exploration is to be done on an individual block.
# With a markdown to explain the Exploration

<br>
<br>

#### Machine Learning

##### Predictive Modeling for Annual Turnover

Using **Annual Turnover** as the target variable (what you want to predict) and use other features as predictors.

* No of Employees
* TCTC
* Basic Salary
* Cash Injection
* Contrib Waiver

we can use regression algorithms like linear regression, decision trees, or random forests to build the model.

<br>

##### Employee Classification

We can use machine learning to classify employees into different categories based

* Cash Injection
* Contrib Waiver
* Affected Employee

We might want to classify employees into **Highly Affected** and **Less Affected** categories.
We can use classification algorithms like logistic regression, decision trees, or support vector machines.

<br>

##### Employee Segmentation

Clustering techniques like K-means clustering can be used to segment employees based on their characteristics.

We can use features like:

* No of Employees
* TCTC
* Basic Salary

To create meaningful clusters

In [20]:
# Code goes here!

<br>
<br>

#### Evaluation of Machine Learning

Present performance metrics (e.g., Mean Absolute Error, R-squared) for each algorithm used.

Explain what the results mean:

* Which algorithm performed better?
* How accurate is the prediction of turnover?

<br>
<br>

#### Presentation of Results

Summarize key findings:

* Trends in employee turnover.
* Compensation fairness insights.

Mention any actionable recommendations based on the analysis.